DQM/BeamMonitor/plugins/BeamSpotDipServer.cc

```diff
@@ -0,0 +1,740 @@
+  }
+
+  list.push_back(line); // remainder
+
+  return list;
+}
+
+/*****************************************************************************/ 
+string BeamSpotDipServer::tkStatus()
+{
+  string outstr;
+
+  if(readFromNFS)
+  { // get from file on /nfs
+    ifstream logfile(sourceFile1);
+
+    if (!logfile.good() || getFileSize(sourceFile1) == 0) {
+      // file does not exist or has zero size
+      outstr = "No CMS Tracker status available. No DAQ/DQM.";
+    } else {
+      int nthLnInRcd = 0;
+      string record;
+
+      try {
+        string record;
+
+        while (getline(logfile,record)) {
+          nthLnInRcd ++;
+          vector<string> tmp = parse(record, " ");
+
+          switch(nthLnInRcd) {
+            case 7  :
+              if (tmp[1].find("Yes") == string::npos)
+                outstr = "CMS Tracker OFF.";
+              else
+                outstr = "CMS not taking data or no beam.";
+              break;
+            case 8  : runnum = stoi(tmp[1]); break;
+            default : break;
+          }
+        }
+      } catch (exception & e) {
+        cerr << "[Dip] Exception (tkStatus): " << e.what() << endl;
+      }
+    }
+
+    logfile.close();
+  }
+  else
+  { 
+    // get from DCS
+    if(wholeTrackerOn) outstr = "CMS not taking data or no beam.";
+                  else outstr = "CMS Tracker OFF.";
+  }
+
+  return outstr;
+}
+
+/*****************************************************************************/ 
+void BeamSpotDipServer::problem()
+{
+  if(verbose) cout << "[Dip] no update | alive = " << alive << endl;
+
+  lsCount++;
+
+  if ((lsCount % timeoutLS[0] == 0)
+   && (lsCount % timeoutLS[1] != 0)) // first time out
+  {
+    if (!alive.test(1)) alive.flip(1);
+    if (!alive.test(2))
+    {
+      if (!alive.test(7)) fakeRcd();
+                     else trueRcd();
+
+      stringstream warnMsg;
+      warnMsg << "No new data for " << lsCount << " LS";
+      publishRcd("Uncertain",warnMsg.str(),false,false);
+    } else {
+      fakeRcd();
+
+      stringstream warnMsg;
+      warnMsg << "No new data for " << lsCount << " LS: "
+               << tkStatus();
+      publishRcd("Bad",warnMsg.str(),false,false);
+    }
+  }
+  else if (lsCount % timeoutLS[1] == 0) // second time out
+  {
+    if (!alive.test(2)) alive.flip(2);
+    fakeRcd();
+
+    stringstream warnMsg;
+    warnMsg << "No new data for " << lsCount << " LS: " << tkStatus();
+    publishRcd("Bad",warnMsg.str(),false,false);
+  }
+}
+
+/*****************************************************************************/ 
+bool BeamSpotDipServer::readRcd(const BeamSpotOnlineObjects & bs)
+// read from database
+{
+  runnum = bs.GetLastAnalyzedRun();
+
+  startTime      = bs.GetCreationTime();
+  startTimeStamp = bs.GetCreationTime(); // not provided by BeamSpotOnlineObject
+    endTime      = bs.GetCreationTime();
+    endTimeStamp = bs.GetCreationTime(); // not provided by BeamSpotOnlineObject
```

Comment:
What is the difference between `starTime` and `startTimeStamp`? (and same for `endTime`).
Maybe they can be added to the BeamSpotOnlineObject.

Reply:
startTime is an std::string, while startTimeStamp is epoch (number of seconds since 1970).
In principle startTime should contain the beginning of the processed lumisection range, and endTime should mark the end of the processed lumisection range.
Can you add those? Actually epoch could be calculated from CreationTime through mktime.

Reply:
The function `GetCreationTime()` for the BeamSpotOnlineObject returns the moment in which the payload was actually created (see [these lines](https://github.com/cms-sw/cmssw/blob/master/DQM/BeamMonitor/plugins/BeamMonitor.cc#L1373-L1375)), it's not related to the LS analyzed.

I'll look into adding these parameters to the online beamspot.